File: LeetCode/Problems/95.go
package main

func main() {

}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func dfs(start, end int) []*TreeNode {
	ret := make([]*TreeNode, 0)
	if start > end {
		return nil
	}
	if start == end {
		return []*TreeNode{{start, nil, nil}}
	}
	for i := start; i <= end; i++ {
		left, right := dfs(start, i-1), dfs(i+1, end)
		for _, l := range left {
			for _, r := range right {
				ret = append(ret, &TreeNode{i, l, r})
			}
		}
	}
	return ret
}

func generateTrees(n int) []*TreeNode {
	return dfs(1, n)
}

File: LeetCode/Problems/95.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
   public:
    vector<TreeNode *> generateTrees(int n) { return genTrees(1, n); }

    vector<TreeNode *> genTrees(int start, int end) {
        vector<TreeNode *> res;
        if (start > end) {
            res.push_back(nullptr);
        } else {
            vector<TreeNode *> left, right;
            for (int i = start; i <= end; i++) {
                left = genTrees(start, i - 1);
                right = genTrees(i + 1, end);
                for (auto lnode : left) {
                    for (auto rnode : right) {
                        TreeNode *root = new TreeNode(i);
                        root->left = lnode;
                        root->right = rnode;
                        res.push_back(root);
                    }
                }
            }
        }
        return res;
    }
};

int main() {
    Solution a;
    a.generateTrees(3);
    return 0;
}